minimint/src/net/peers.rs

```diff
@@ -0,0 +1,457 @@
+        match target {
+            Target::All => {
+                for connection in self.connections.values_mut() {
+                    connection.send(msg.clone()).await;
+                }
+            }
+            Target::Node(peer_id) => {
+                if let Some(peer) = self.connections.get_mut(&peer_id) {
+                    peer.send(msg).await;
+                } else {
+                    trace!(peer = ?peer_id, "Not sending message to unknown peer (maybe banned)");
+                }
+            }
+        }
+    }
+
+    async fn receive(&mut self) -> (PeerId, T) {
+        // TODO: optimize, don't throw away remaining futures
+
+        let futures_non_banned = self.connections.iter_mut().map(|(&peer, connection)| {
+            let receive_future = async move {
+                let msg = connection.receive().await;
+                (peer, msg)
+            };
+            Box::pin(receive_future)
+        });
+
+        select_all(futures_non_banned).await.0
+    }
+
+    async fn ban_peer(&mut self, peer: PeerId) {
+        self.connections.remove(&peer);
+        warn!("Peer {} banned.", peer);
+    }
+}
+
+impl<M> PeerConnection<M>
+where
+    M: Debug + Clone + Send + Sync + 'static,
+{
+    fn new(
+        id: PeerId,
+        cfg: ConnectionConfig,
+        connect: SharedAnyConnector<PeerMessage<M>>,
+        incoming_connections: Receiver<AnyFramedTransport<PeerMessage<M>>>,
+    ) -> PeerConnection<M> {
+        let (outgoing_sender, outgoing_receiver) = tokio::sync::mpsc::channel::<M>(1024);
+        let (incoming_sender, incoming_receiver) = tokio::sync::mpsc::channel::<M>(1024);
+
+        let io_thread = tokio::spawn(Self::run_io_thread(
+            incoming_sender,
+            outgoing_receiver,
+            id,
+            cfg,
+            connect,
+            incoming_connections,
+        ));
+
+        PeerConnection {
+            outgoing: outgoing_sender,
+            incoming: incoming_receiver,
+            _io_task: io_thread,
+        }
+    }
+
+    async fn send(&mut self, msg: M) {
+        self.outgoing.send(msg).await.expect("io task died");
+    }
+
+    async fn receive(&mut self) -> M {
+        self.incoming.recv().await.expect("io task died")
+    }
+
+    async fn run_io_thread(
+        incomming: Sender<M>,
+        mut outgoing: Receiver<M>,
+        peer: PeerId,
+        cfg: ConnectionConfig,
+        connect: SharedAnyConnector<PeerMessage<M>>,
+        mut incoming_connections: Receiver<AnyFramedTransport<PeerMessage<M>>>,
+    ) {
+        let mut connection: Option<AnyFramedTransport<PeerMessage<M>>> = None;
+        let mut failed_reconnect_counter: u64 = 1;
+        let mut reconnect: Option<Instant> = Some(reconnect_time(failed_reconnect_counter));
+        let mut resend_queue = MessageQueue::default();
+        let mut last_received: Option<MessageId> = None;
+
+        let mut err = None;
+        loop {
+            if let Some(e) = err.take() {
```

Comment:
At some point I'd like to model that whole state machine in a nicer way, maybe with an enum? Idk yet.

Reply:
Yes, enums are very good for that. It also helps understand code better - explaining all possible states before the logic